src/NetMQ/ReceiveThreadSafeSocketExtensions.cs

```diff
@@ -88,15 +101,20 @@ public static bool TryReceiveBytes([NotNull] this IThreadSafeInSocket socket, ou
         /// Receive a bytes from <paramref name="socket"/> asynchronously.
         /// </summary>
         /// <param name="socket">The socket to receive from.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="CancellationToken.None"/>.</param>
         /// <returns>The content of the received message.</returns>
-        public static ValueTask<byte[]> ReceiveBytesAsync([NotNull] this IThreadSafeInSocket socket)
+        /// <exception cref="System.OperationCanceledException">The token has had cancellation requested.</exception>
+        public static ValueTask<byte[]> ReceiveBytesAsync([NotNull] this IThreadSafeInSocket socket,
+            CancellationToken cancellationToken = default)
         {
             if (TryReceiveBytes(socket, out var bytes))
                 return new ValueTask<byte[]>(bytes);
 
             // TODO: this is a hack, eventually we need kind of IO ThreadPool for thread-safe socket to wait on asynchronously
             // and probably implement IValueTaskSource
-            return new ValueTask<byte[]>(Task.Factory.StartNew(socket.ReceiveBytes, TaskCreationOptions.LongRunning));
+            // TODO: should we avoid lambda here as it cause heap allocation for the environment?
+            return new ValueTask<byte[]>(Task.Factory.StartNew(() => socket.ReceiveBytes(cancellationToken),
```

Comment:
@drewnoakes do you think we would benefit here from de-lambda the action? using a real method with the cancellation token passed as a state object?

Reply:
It would avoid a heap allocation for the closure object, yes. You'd need to pass both the socket and token as state, then cast them at the other side.

I played around with this a little:

- [lambda](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABADAAgwRgDoAVAC1gEMATASwDsBzAbgFgAobPfAVlY4GY8AJhwBhdgG92OGXkEA1CgBsArjAxIAPASwA+HAFkAFMAgQlOYABoxFOmBhKlFAC40IdYhADWMOjhcfPwBKaVkpNlkonBoAMxwTUMjolIwAdhw6GAB3HEVVdS0dXSMAIhcoNVLgvhSZMNSMrNz8tQ1tfD0jDUIAMQowQKgAT0IAZRcKKBcAORyjI2CcAF59QN86QgBJAGdROwcnV3c6ACUYAEc1HZcYKhwAfhxSsAPHJTvSnBBn2OUdmDVGoNHAAX3YoKAA==)
- [method with state passing](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABADAAgwRgDoAVAC1gEMATASwDsBzAbgFgAobPfAVlY4GY8AJhwBhdgG92OGXkEA1CgBsArjAxIAPASwA+HAFkAFMAgQlOYABoxFOmBhKlFAC40IdYhADWMOjhcfPwBKaVkpNlkonBoAMxwTUMjolIwAdhw6GAB3HEVVdS0dXSMAIhcoNVLgvhSZMNSMrNz8tQ1tfD0jDUIAMQowQKgAT0IAZRcKKBcAORyjXptA3zoloLpgmoacAF9t7YAHKBoAN1cYLlxeowhgACsYQYD1pKiIurwMoyNROwcnVzuTwvZYhQgASQAzr97I5nG4PAAlGAARzUkJcMCoOAA/DhSmA/nCsaUcCB8bFlJCYKVarI9mwdkA==)
- [local function with state passing](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABADAAgwRgDoAVAC1gEMATASwDsBzAbgFgAobPfAVlY4GY8AJhwBhdgG92OGXkEA1CgBsArjAxIAPASwA+HAFkAFMAgQlOYABoxFOmBhKlFAC40IdYhADWMOjhcfPwBKaVkpNlkonBoAMxwTUMjolIwAdhw6GAB3HEVVdS0dXSMAIhcoNVLgvhSZMNSMrNz8tQ1tfD0jDUIAMQowQKgAT0IAZRcKKBcAORyjXptA3zoloLpgmobo7aiCJC5cXqMIYAArGEGA9aS6nAi7vYyjI1E7BydXd08b5ZDCACSAGc3vZHM43B4AEowACOaiBLhgVBwAH4cKUwO9wcjSjgQBjYsogTBSrU6gBfbZUtgUoA)

It'd be interesting to use BenchmarkDotNet to get some data on allocations for send/receive in various scenarios.

Reply:
If I have to wrap both the socket and the token I'm not sure I have a benefit of using state instead of lambda.

Reply:
With extension method seems interesting:

https://sharplab.io/#gist:e7e76a5ece9d763149b8fb1eeed2587b

The question is, what is __ldftn is doing...

Reply:
seems to be allocation free:

https://docs.microsoft.com/en-us/dotnet/api/system.reflection.emit.opcodes.ldftn?view=netcore-3.1

Reply:
solved it with private extension method.

Reply:
Actually casting `CancellationToken` as `object` will box, which allocates. I'm not sure there's a way around allocating here. I'd wondered about making the method `async ValueTask<>` but didn't test it.

Reply:
The main thing is that these methods don't allocate when they complete synchronously. If they need to yield, `ValueTask<>` will allocate anyway.

Reply:
Yes, you are right. I will simplify it back and merge it.